examples/linear_model/plot_lasso_and_elasticnet.py

```diff
@@ -1,97 +1,245 @@
+# signals. Only the 10 lowest out of the 100 frequencies in `X` are used to
+# generate `y`, while the rest of the features are not informative. This results
+# in a high dimentional sparse feature space, where some degree of
+# l1-penalization is necessary.
 
 import numpy as np
-import matplotlib.pyplot as plt
-
-from sklearn.metrics import r2_score
 
-np.random.seed(42)
+rng = np.random.RandomState(0)
+n_samples, n_features, n_informative = 50, 100, 10
+time_step = np.linspace(-2, 2, n_samples)
+freqs = 2 * np.pi * np.sort(rng.rand(n_features)) / 0.01
+X = np.zeros((n_samples, n_features))
 
-n_samples, n_features = 50, 100
-X = np.random.randn(n_samples, n_features)
+for i in range(n_features):
+    X[:, i] = np.sin(freqs[i] * time_step)
 
-# Decreasing coef w. alternated signs for visualization
 idx = np.arange(n_features)
-coef = (-1) ** idx * np.exp(-idx / 10)
-coef[10:] = 0  # sparsify coef
-y = np.dot(X, coef)
+true_coef = (-1) ** idx * np.exp(-idx / 10)
+true_coef[n_informative:] = 0  # sparsify coef
+y = np.dot(X, true_coef)
 
-# Add noise
-y += 0.01 * np.random.normal(size=n_samples)
+# %%
+# Some of the informative features have close frequencies to induce
+# (anti-)correlations.
 
-# Split data in train set and test set
-n_samples = X.shape[0]
-X_train, y_train = X[: n_samples // 2], y[: n_samples // 2]
-X_test, y_test = X[n_samples // 2 :], y[n_samples // 2 :]
+freqs[:n_informative]
 
 # %%
-# Lasso
-# ---------------------------------------------------
+# A random phase is introduced using `numpy.random.random_sample
+# <https://numpy.org/doc/stable/reference/random/generated/numpy.random.random_sample.html>`_
+# and some gaussian noise (implemented by `numpy.random.normal
+# <https://numpy.org/doc/stable/reference/random/generated/numpy.random.random_sample.html>`_)
+# is added to both the features and the target.
+
+for i in range(n_features):
+    X[:, i] = np.sin(freqs[i] * time_step + 2 * (rng.random_sample() - 0.5))
+    X[:, i] += 0.2 * rng.normal(0, 1, n_samples)
+
+y += 0.2 * rng.normal(0, 1, n_samples)
+
+# %%
+# Such sparse, noisy and correlated features can be obtained, for instance, from
+# sensor nodes monitoring some environmental variables, as they typically register
+# similar values depending on their positions (spatial correlations).
+# We can visualize the target.
+
+import matplotlib.pyplot as plt
+
+plt.plot(time_step, y)
+plt.ylabel("target signal")
+plt.xlabel("time")
+_ = plt.title("Superposition of sinusoidal signals")
 
+# %%
+# We split the data into train and test sets for simplicity. In practice one should
+# use a :class:`~sklearn.model_selection.TimeSeriesSplit` cross-validation to
+# estimate the variance of the test score. Here we set `shuffle="False"` to keep
+# the causality of the unseen data.
+
+from sklearn.model_selection import train_test_split
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X,
+    y,
+    test_size=0.5,
+    shuffle=False,
+)
+
+# %%
```

Comment:
Maybe we can introduce a bit what will come later since we don't discuss the results right after each cell.
So we can mention that we will check the time and statistical performance in terms of goodness of fit (i.e. R2) and we will make a final comparison of all models.